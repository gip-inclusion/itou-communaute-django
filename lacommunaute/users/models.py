import hashlib
from uuid import uuid4

from dateutil.relativedelta import relativedelta
from django.conf import settings
from django.contrib.auth.models import AbstractUser, UserManager as BaseUserManager
from django.db import models
from django.utils import timezone

from lacommunaute.users.enums import EmailLastSeenKind, IdentityProvider


class UserManager(BaseUserManager):
    def create_user(self, username=None, email=None, password=None, **extra_fields):
        if not username:
            username = str(uuid4())
        return super().create_user(username, email, password, **extra_fields)


class User(AbstractUser):
    identity_provider = models.CharField(
        max_length=2,
        choices=IdentityProvider.choices,
        null=False,
        blank=False,
    )

    objects = UserManager()

    def __str__(self):
        return self.email


class EmailLastSeenQuerySet(models.QuerySet):
    def seen(self, email, kind):
        if kind not in [kind for kind, _ in EmailLastSeenKind.choices]:
            raise ValueError(f"Invalid kind: {kind}")

        return EmailLastSeen.objects.bulk_create(
            [EmailLastSeen(email=email, last_seen_at=timezone.now(), last_seen_kind=kind, missyou_send_at=None)],
            update_fields=["last_seen_at", "last_seen_kind", "missyou_send_at"],
            update_conflicts=True,
            unique_fields=["email"],
        )

    def eligible_to_missyou_message(self):
        return self.filter(
            last_seen_at__lte=timezone.now() - relativedelta(months=settings.EMAIL_LAST_SEEN_MISSYOU_DELAY),
            missyou_send_at=None,
        ).order_by("last_seen_at")

    def eligible_to_soft_deletion(self):
        return self.filter(
            missyou_send_at__lte=timezone.now()
            - relativedelta(days=settings.EMAIL_LAST_SEEN_ARCHIVE_PERSONNAL_DATAS_DELAY),
            deleted_at=None,
        )


class EmailLastSeen(models.Model):
    email = models.EmailField(verbose_name="email", null=False, unique=True)
    email_hash = models.CharField(max_length=255, verbose_name="email hash", null=True)
    last_seen_at = models.DateTimeField(verbose_name="last seen at", null=False)
    last_seen_kind = models.CharField(
        max_length=12, verbose_name="last seen kind", choices=EmailLastSeenKind.choices, null=False
    )
    missyou_send_at = models.DateTimeField(verbose_name="miss you sent at", null=True, blank=True)
    deleted_at = models.DateTimeField(verbose_name="deleted at", null=True, blank=True)

    objects = EmailLastSeenQuerySet.as_manager()

    def soft_delete(self):
        self.deleted_at = timezone.now()
        salted_email = f"{self.email}-{settings.EMAIL_LAST_SEEN_HASH_SALT}"
        self.email_hash = hashlib.sha256(salted_email.encode("utf-8")).hexdigest()
        self.email = f"email-anonymis√©-{self.pk}"
        self.save()
