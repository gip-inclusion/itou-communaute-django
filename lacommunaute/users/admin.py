from django.contrib import admin
from django.contrib.auth.admin import UserAdmin
from django.contrib.auth.models import Group

from lacommunaute.users.models import User


class UserAdmin(UserAdmin):
    list_display = ("username", "email", "first_name", "last_name", "identity_provider", "is_staff")
    list_filter = ("is_staff", "is_superuser", "is_active", "identity_provider", "groups")


admin.site.register(User, UserAdmin)


class GroupUserInline(admin.TabularInline):
    model = Group.user_set.through
    list_display = "user_set__email"
    raw_id_fields = ("user",)
    extra = 1


class GroupAdmin(admin.ModelAdmin):
    inlines = [GroupUserInline]


admin.site.unregister(Group)
admin.site.register(Group, GroupAdmin)
