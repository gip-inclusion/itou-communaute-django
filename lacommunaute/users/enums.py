from django.db import models


class IdentityProvider(models.TextChoices):
    INCLUSION_CONNECT = "IC", "Inclusion Connect"
    PRO_CONNECT = "PC", "Pro Connect"
