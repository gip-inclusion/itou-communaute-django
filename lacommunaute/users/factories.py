import functools
import random

import factory
from django.contrib.auth.hashers import make_password
from django.contrib.auth.models import Group
from machina.core.db.models import get_model


User = get_model("users", "User")

DEFAULT_PASSWORD = "supercalifragilisticexpialidocious"


@functools.cache
def default_password():
    return make_password(DEFAULT_PASSWORD)


class GroupFactory(factory.django.DjangoModelFactory):
    class Meta:
        model = Group


class UserFactory(factory.django.DjangoModelFactory):
    class Meta:
        model = User

    username = factory.LazyAttribute(
        lambda a: "{}_{}_{}".format(a.first_name, a.last_name, random.randrange(111, 999))
    )
    first_name = factory.Faker("first_name")
    last_name = factory.Faker("last_name")
    email = factory.Faker("email")
    password = factory.LazyFunction(default_password)

    @factory.post_generation
    def with_perm(obj, create, extracted, **kwargs):
        if not create or not extracted:
            return
        obj.user_permissions.add(*extracted)
