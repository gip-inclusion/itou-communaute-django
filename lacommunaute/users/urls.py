from django.urls import path

from lacommunaute.users.views import (
    CreateUserView,
    LoginView,
    login_with_link,
    logout_view,
)


app_name = "users"

urlpatterns = [
    path("create/", CreateUserView.as_view(), name="create"),
    path("login/", LoginView.as_view(), name="login"),
    path("login/<str:uidb64>/<str:token>/", login_with_link, name="login_with_link"),
    path("logout/", logout_view, name="logout"),
]
