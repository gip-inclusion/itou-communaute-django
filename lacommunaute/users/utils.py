import hashlib

from django.conf import settings
from django.db import transaction
from django.utils import timezone

from lacommunaute.forum_conversation.models import Post
from lacommunaute.users.models import EmailLastSeen, User


@transaction.atomic
def soft_delete_users(email_last_seen_list):
    emails = [email_last_seen.email for email_last_seen in email_last_seen_list]

    users = list(User.objects.filter(email__in=emails))
    posts = list(Post.objects.filter(username__in=emails))

    anonymized_emails = {}

    for email_last_seen in email_last_seen_list:
        email_last_seen.deleted_at = timezone.now()
        salted_email = f"{email_last_seen.email}-{settings.EMAIL_LAST_SEEN_HASH_SALT}"
        email_last_seen.email_hash = hashlib.sha256(salted_email.encode("utf-8")).hexdigest()
        anonymized_emails[email_last_seen.email] = f"email-anonymise-{email_last_seen.id}@{settings.COMMU_FQDN}"
        email_last_seen.email = anonymized_emails[email_last_seen.email]

    for user in users:
        user.email = anonymized_emails[user.email]
        user.first_name = "Anonyme"
        user.last_name = "Anonyme"

    for post in posts:
        post.username = anonymized_emails[post.username]

    user_count = User.objects.bulk_update(users, ["email", "first_name", "last_name"])
    post_count = Post.objects.bulk_update(posts, ["username"])
    email_last_seen_count = EmailLastSeen.objects.bulk_update(
        email_last_seen_list,
        ["email", "email_hash", "deleted_at"],
    )

    return user_count, post_count, email_last_seen_count
