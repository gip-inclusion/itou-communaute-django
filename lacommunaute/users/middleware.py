from logging import getLogger

from django.utils.deprecation import MiddlewareMixin

from lacommunaute.users.enums import EmailLastSeenKind
from lacommunaute.users.models import EmailLastSeen


logger = getLogger("lacommunaute")


class MarkAsSeenLoggedUserMiddleware(MiddlewareMixin):
    def process_request(self, request):
        if request.user.is_authenticated:
            EmailLastSeen.objects.seen(email=request.user.email, kind=EmailLastSeenKind.LOGGED)
