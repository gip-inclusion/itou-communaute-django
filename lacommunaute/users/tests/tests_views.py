import json
import re
from unittest import mock
from urllib.parse import urlencode

import httpx
import pytest
import respx
from django.conf import settings
from django.contrib.auth.tokens import default_token_generator
from django.contrib.messages.api import get_messages
from django.contrib.messages.middleware import MessageMiddleware
from django.contrib.sessions.middleware import SessionMiddleware
from django.test import RequestFactory, override_settings
from django.urls import reverse
from django.utils.encoding import force_bytes
from django.utils.http import urlsafe_base64_encode
from pytest_django.asserts import assertContains

from lacommunaute.forum_member.models import ForumProfile
from lacommunaute.notification.emails import SIB_SMTP_URL
from lacommunaute.users.enums import IdentityProvider
from lacommunaute.users.factories import UserFactory
from lacommunaute.users.models import User
from lacommunaute.users.views import send_magic_link
from lacommunaute.utils.enums import Environment
from lacommunaute.utils.testing import parse_response_to_soup
from lacommunaute.utils.urls import clean_next_url


next_url_tuples = [("/", "/"), ("/topics/", "/topics/"), ("http://www.unallowed_host.com", "/")]


@pytest.fixture(name="mock_respx_post_to_sib_smtp_url")
def mock_respx_post_to_sib_smtp_url_fixture():
    with respx.mock:
        respx.post(SIB_SMTP_URL).mock(return_value=httpx.Response(200, json={"message": "OK"}))
        yield


@pytest.fixture(name="mock_token_generator")
def mock_token_generator_fixture():
    with mock.patch.object(default_token_generator, "make_token", return_value="fixed-token"):
        yield default_token_generator


def custom_user_token(user, fixed_token=False):
    uidb64 = urlsafe_base64_encode(force_bytes(user.pk))
    if fixed_token:
        token = "fixed-token"
    else:
        token = default_token_generator.make_token(user)
    return user, uidb64, token


@pytest.fixture(name="user")
def user_fixture():
    return UserFactory(
        first_name="Samuel",
        last_name="Jackson",
        email="samuel@jackson.com",
        identity_provider=IdentityProvider.PRO_CONNECT.value,
    )


@pytest.fixture(name="user_token")
def user_token_fixture(user):
    return custom_user_token(user)


@pytest.fixture(name="fixed_user_token")
def fixed_user_token_fixture(user):
    return custom_user_token(user, fixed_token=True)


def validate_magiclink_payload(payload_as_str, uidb64, token, expected):
    url = f"{settings.COMMU_PROTOCOL}://{settings.COMMU_FQDN}" + reverse(
        "users:login_with_link", kwargs={"uidb64": uidb64, "token": token}
    )
    payload = json.loads(payload_as_str)

    if payload["params"]["login_link"] == url + "?" + urlencode({"next": expected}):
        return True

    return False


class TestSendMagicLink:
    @pytest.mark.parametrize("env,count_msg", [(Environment.PROD, 0), (Environment.DEV, 1)])
    def test_send_magic_link(
        self, db, user, snapshot, mock_token_generator, mock_respx_post_to_sib_smtp_url, env, count_msg
    ):
        with override_settings(ENVIRONMENT=env):
            next_url = "/topics/"
            request = RequestFactory().get("/")
            SessionMiddleware(lambda request: None).process_request(request)
            MessageMiddleware(lambda request: None).process_request(request)
            send_magic_link(request, user, next_url)

            token = default_token_generator.make_token(user)
            uidb64 = urlsafe_base64_encode(force_bytes(user.pk))
            url = reverse("users:login_with_link", kwargs={"uidb64": uidb64, "token": token})
            query_params = urlencode({"next": clean_next_url(next_url)})
            login_link = f"{settings.COMMU_PROTOCOL}://{settings.COMMU_FQDN}{url}?{query_params}"

            payload_as_str = respx.calls[0].request.content.decode()
            payload = json.loads(payload_as_str)
            assert payload["params"]["login_link"] == login_link
            assert payload_as_str.replace(login_link, "LOGIN_LINK") == snapshot(name="send_magic_link_payload")

            # we want messages do not appear in the productive environment
            msgs = get_messages(request)
            assert len(msgs._queued_messages) == count_msg
            if msgs._queued_messages:
                assert str(msgs._queued_messages[0]) == f'<a href="{login_link}">{login_link}</a> sent to {user.email}'


class TestLoginView:
    @pytest.mark.parametrize("next_url", [None, "/", "/topics/", "http://www.unallowed_host.com"])
    def test_content(self, client, db, next_url, snapshot):
        url = reverse("users:login") + f"?next={next_url}" if next_url else reverse("users:login")
        response = client.get(url)
        assert response.status_code == 200
        content = parse_response_to_soup(response, selector="main")
        assert str(content) == snapshot(name="login_view_content")

    @override_settings(ENVIRONMENT=Environment.PROD)
    @pytest.mark.parametrize("next_url,expected", next_url_tuples)
    def test_post(
        self,
        client,
        db,
        fixed_user_token,
        next_url,
        expected,
        snapshot,
        mock_token_generator,
        mock_respx_post_to_sib_smtp_url,
    ):
        user, uidb64, token = fixed_user_token
        response = client.post(
            reverse("users:login"),
            data={"email": user.email, "next": next_url},
        )

        assert response.status_code == 200
        content = parse_response_to_soup(response, selector="main")
        assert str(content) == snapshot(name="login_view_content")

        payload_as_str = respx.calls[0].request.content.decode()
        assert validate_magiclink_payload(payload_as_str, uidb64, token, expected)

    @pytest.mark.parametrize("next_url,expected", next_url_tuples)
    def test_redirection_when_email_is_unknown(self, client, db, next_url, expected, mock_respx_post_to_sib_smtp_url):
        response = client.post(
            reverse("users:login"),
            data={"email": "john@travolta.es", "next": next_url},
        )
        assert response.status_code == 302
        assert response.url == reverse("users:create") + "?" + urlencode(
            {"email": "john@travolta.es", "next": expected}
        )

    @pytest.mark.parametrize("next_url,expected", next_url_tuples)
    def test_user_is_already_authenticated(self, client, db, next_url, expected):
        user = UserFactory()
        client.force_login(user)
        url = reverse("users:login") + f"?next={next_url}"
        response = client.get(url)
        assert response.status_code == 302
        assert response.url == expected


class TestCreateUserView:
    @override_settings(ENVIRONMENT=Environment.PROD)
    @pytest.mark.parametrize("next_url,expected", next_url_tuples)
    def test_post_new_email(
        self, client, db, next_url, expected, snapshot, mock_token_generator, mock_respx_post_to_sib_smtp_url
    ):
        email = "john@travolta.lt"
        response = client.post(
            reverse("users:create"),
            data={"email": email, "first_name": "John", "last_name": "Travolta", "next": next_url},
        )
        assert response.status_code == 200

        content = parse_response_to_soup(response, selector="main")
        assert str(content) == snapshot(name="create_user_view_content")

        created_user = User.objects.get(email=email)
        pattern = r"^[a-f0-9]{8}(-[a-f0-9]{4}){3}-[a-f0-9]{12}$"
        assert re.match(pattern, created_user.username)
        assert created_user.email == email
        assert created_user.first_name == "John"
        assert created_user.last_name == "Travolta"
        assert created_user.identity_provider == IdentityProvider.MAGIC_LINK
        created_forum_profile = ForumProfile.objects.get(user=created_user)
        assert created_forum_profile.user.email == email

        _, uidb64, token = custom_user_token(created_user, fixed_token=True)
        payload_as_str = respx.calls[0].request.content.decode()
        assert validate_magiclink_payload(payload_as_str, uidb64, token, expected)

    @override_settings(ENVIRONMENT=Environment.PROD)
    @pytest.mark.parametrize("next_url,expected", next_url_tuples)
    def test_post_existing_email(
        self, client, db, user, next_url, expected, snapshot, mock_token_generator, mock_respx_post_to_sib_smtp_url
    ):
        first_name = user.first_name
        last_name = user.last_name
        response = client.post(
            reverse("users:create"),
            data={"email": user.email, "first_name": "Adam", "last_name": "Smith", "next": next_url},
        )
        assert response.status_code == 200
        content = parse_response_to_soup(response, selector="main")
        assert str(content) == snapshot(name="create_user_view_content")

        user.refresh_from_db()
        assert user.first_name == first_name
        assert user.last_name == last_name
        assert user.identity_provider == IdentityProvider.PRO_CONNECT

        _, uidb64, token = custom_user_token(user, fixed_token=True)
        payload_as_str = respx.calls[0].request.content.decode()
        assert validate_magiclink_payload(payload_as_str, uidb64, token, expected)


class TestLoginWithLinkView:
    def test_user_not_found(self, client, db):
        response = client.get(
            reverse("users:login_with_link", kwargs={"uidb64": "dGVzdEB0ZXN0LmNvbQ==", "token": "token"})
        )
        assert response.status_code == 302
        assert response.url == reverse("users:login")

    def test_invalid_token(self, client, db, user_token):
        _, uidb64, __doc__ = user_token
        response = client.get(reverse("users:login_with_link", kwargs={"uidb64": uidb64, "token": "invalid_token"}))
        assert response.status_code == 302
        assert response.url == reverse("users:login")

    def test_expired_token(self, client, db, user_token):
        user, uidb64, token = user_token

        client.force_login(user)
        # force logout
        client.session.flush()

        response = client.get(reverse("users:login_with_link", kwargs={"uidb64": uidb64, "token": token}))
        assert response.status_code == 302
        assert response.url == reverse("users:login")

    def test_success(self, client, db, user_token):
        _, uidb64, token = user_token
        next_url = "/topics/"

        response = client.get(
            reverse("users:login_with_link", kwargs={"uidb64": uidb64, "token": token}), data={"next": next_url}
        )
        assert response.status_code == 302
        assert response.url == next_url
        assert client.session.get(IdentityProvider.MAGIC_LINK.name) == 1

    def test_unallowed_next_url(self, client, db, user_token):
        _, uidb64, token = user_token
        next_url = "http://www.unallowed_host.com"

        response = client.get(
            reverse("users:login_with_link", kwargs={"uidb64": uidb64, "token": token}), data={"next": next_url}
        )
        assert response.status_code == 302
        assert response.url == "/"


class TestLogoutView:
    @pytest.mark.parametrize(
        "login,expected_logout_url",
        [("proconnect", reverse("openid_connect:logout")), ("magiclink", reverse("users:logout"))],
    )
    def test_logout_url_in_template(self, client, db, login, expected_logout_url):
        user = UserFactory()
        client.force_login(user)

        if login == "magiclink":
            session = client.session
            session[IdentityProvider.MAGIC_LINK.name] = 1
            session.save()

        response = client.get(reverse("pages:home"))
        assertContains(response, expected_logout_url)

    def test_logout_from_magiclink_session(self, client, db):
        user = UserFactory()
        client.force_login(user)
        response = client.get(reverse("users:logout"))
        assert response.status_code == 302
        assert response.url == reverse("pages:home")
        assert client.session.get("_auth_user_id") is None
        assert client.session.get("_auth_user_backend") is None
