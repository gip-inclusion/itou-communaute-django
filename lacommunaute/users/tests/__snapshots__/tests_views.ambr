# serializer version: 1
# name: TestCreateUserView.test_post_existing_email[/-/][create_user_view_content]
  '''
  <main class="s-main" id="main" role="main">
              
                  
              
              
              
                  
      <section class="s-title-01 mt-lg-5">
          <div class="s-title-01__container container">
              <div class="s-title-01__row row">
                  <div class="s-title-01__col col-lg-8 col-12">
                      <h1 class="s-title-01__title h1">Se connecter | S'inscrire</h1>
                  </div>
              </div>
          </div>
      </section>
      <section class="s-section">
          <div class="s-section__container container">
              <div class="s-section__row row">
                  <div class="s-section__col col-12 col-lg-7">
                      <p>
                          Un lien de connexion vous a été envoyé à l'adresse samuel@jackson.com. Veuillez vérifier votre boîte de réception et cliquer sur le lien pour vous connecter.
                      </p>
                  </div>
              </div>
          </div>
      </section>
  
              
          </main>
  '''
# ---
# name: TestCreateUserView.test_post_existing_email[/topics/-/topics/][create_user_view_content]
  '''
  <main class="s-main" id="main" role="main">
              
                  
              
              
              
                  
      <section class="s-title-01 mt-lg-5">
          <div class="s-title-01__container container">
              <div class="s-title-01__row row">
                  <div class="s-title-01__col col-lg-8 col-12">
                      <h1 class="s-title-01__title h1">Se connecter | S'inscrire</h1>
                  </div>
              </div>
          </div>
      </section>
      <section class="s-section">
          <div class="s-section__container container">
              <div class="s-section__row row">
                  <div class="s-section__col col-12 col-lg-7">
                      <p>
                          Un lien de connexion vous a été envoyé à l'adresse samuel@jackson.com. Veuillez vérifier votre boîte de réception et cliquer sur le lien pour vous connecter.
                      </p>
                  </div>
              </div>
          </div>
      </section>
  
              
          </main>
  '''
# ---
# name: TestCreateUserView.test_post_existing_email[http://www.unallowed_host.com-/][create_user_view_content]
  '''
  <main class="s-main" id="main" role="main">
              
                  
              
              
              
                  
      <section class="s-title-01 mt-lg-5">
          <div class="s-title-01__container container">
              <div class="s-title-01__row row">
                  <div class="s-title-01__col col-lg-8 col-12">
                      <h1 class="s-title-01__title h1">Se connecter | S'inscrire</h1>
                  </div>
              </div>
          </div>
      </section>
      <section class="s-section">
          <div class="s-section__container container">
              <div class="s-section__row row">
                  <div class="s-section__col col-12 col-lg-7">
                      <p>
                          Un lien de connexion vous a été envoyé à l'adresse samuel@jackson.com. Veuillez vérifier votre boîte de réception et cliquer sur le lien pour vous connecter.
                      </p>
                  </div>
              </div>
          </div>
      </section>
  
              
          </main>
  '''
# ---
# name: TestCreateUserView.test_post_new_email[/-/][create_user_view_content]
  '''
  <main class="s-main" id="main" role="main">
              
                  
              
              
              
                  
      <section class="s-title-01 mt-lg-5">
          <div class="s-title-01__container container">
              <div class="s-title-01__row row">
                  <div class="s-title-01__col col-lg-8 col-12">
                      <h1 class="s-title-01__title h1">Se connecter | S'inscrire</h1>
                  </div>
              </div>
          </div>
      </section>
      <section class="s-section">
          <div class="s-section__container container">
              <div class="s-section__row row">
                  <div class="s-section__col col-12 col-lg-7">
                      <p>
                          Un lien de connexion vous a été envoyé à l'adresse john@travolta.lt. Veuillez vérifier votre boîte de réception et cliquer sur le lien pour vous connecter.
                      </p>
                  </div>
              </div>
          </div>
      </section>
  
              
          </main>
  '''
# ---
# name: TestCreateUserView.test_post_new_email[/topics/-/topics/][create_user_view_content]
  '''
  <main class="s-main" id="main" role="main">
              
                  
              
              
              
                  
      <section class="s-title-01 mt-lg-5">
          <div class="s-title-01__container container">
              <div class="s-title-01__row row">
                  <div class="s-title-01__col col-lg-8 col-12">
                      <h1 class="s-title-01__title h1">Se connecter | S'inscrire</h1>
                  </div>
              </div>
          </div>
      </section>
      <section class="s-section">
          <div class="s-section__container container">
              <div class="s-section__row row">
                  <div class="s-section__col col-12 col-lg-7">
                      <p>
                          Un lien de connexion vous a été envoyé à l'adresse john@travolta.lt. Veuillez vérifier votre boîte de réception et cliquer sur le lien pour vous connecter.
                      </p>
                  </div>
              </div>
          </div>
      </section>
  
              
          </main>
  '''
# ---
# name: TestCreateUserView.test_post_new_email[http://www.unallowed_host.com-/][create_user_view_content]
  '''
  <main class="s-main" id="main" role="main">
              
                  
              
              
              
                  
      <section class="s-title-01 mt-lg-5">
          <div class="s-title-01__container container">
              <div class="s-title-01__row row">
                  <div class="s-title-01__col col-lg-8 col-12">
                      <h1 class="s-title-01__title h1">Se connecter | S'inscrire</h1>
                  </div>
              </div>
          </div>
      </section>
      <section class="s-section">
          <div class="s-section__container container">
              <div class="s-section__row row">
                  <div class="s-section__col col-12 col-lg-7">
                      <p>
                          Un lien de connexion vous a été envoyé à l'adresse john@travolta.lt. Veuillez vérifier votre boîte de réception et cliquer sur le lien pour vous connecter.
                      </p>
                  </div>
              </div>
          </div>
      </section>
  
              
          </main>
  '''
# ---
# name: TestLoginView.test_content[/][login_view_content]
  '''
  <main class="s-main" id="main" role="main">
              
                  
              
              
              
                  
      <section class="s-title-01 mt-lg-5">
          <div class="s-title-01__container container">
              <div class="s-title-01__row row">
                  <div class="s-title-01__col col-lg-8 col-12">
                      <h1 class="s-title-01__title h1">Se connecter | S'inscrire</h1>
                  </div>
              </div>
          </div>
      </section>
      <section class="s-section">
          <div class="s-section__container container">
              <div class="s-section__row row">
                  <div class="s-section__col col-12 col-lg-7">
                      <div class="c-form">
                          <div class="text-center">
                              <a class="proconnect-button" href="/pro_connect/authorize?next="></a>
                              <p>
                                  <a href="https://proconnect.gouv.fr/" rel="noopener noreferrer" target="_blank" title="Qu’est-ce que AgentConnect ? - nouvelle fenêtre">
                                      Qu’est-ce que ProConnect ?
                                  </a>
                              </p>
                          </div>
                          <hr class="my-5" data-it-text="ou"/>
                          <form action="." enctype="multipart/form-data" method="post" novalidate="">
                              <input name="csrfmiddlewaretoken" type="hidden" value="NORMALIZED_CSRF_TOKEN"/>
                              
                              <fieldset>
                                  <p class="h4">Se connecter avec votre email</p>
                                  <p class="text-muted">✨ Nous allons vous envoyer un code magique pour vous connecter sans mot de passe.</p>
                                  
  
  <div class="form-group" id="div_id_email">
      
          
          
              <input class="form-control" id="id_email" maxlength="320" name="email" placeholder="Votre adresse email" required="" type="email"/>
          
      
      
      
      
  </div>
  
                              </fieldset>
                              <div class="row">
                                  <div class="col-12">
                                      <div class="form-row align-items-center justify-content-end gx-3">
                                          <div class="form-group col col-lg-auto order-2 order-lg-3">
                                              <input name="next" type="hidden" value="/"/>
                                              <input class="btn btn-block btn-primary" type="submit" value="Recevoir le lien de connexion"/>
                                          </div>
                                      </div>
                                  </div>
                              </div>
                          </form>
                      </div>
                  </div>
              </div>
          </div>
      </section>
  
              
          </main>
  '''
# ---
# name: TestLoginView.test_content[/topics/][login_view_content]
  '''
  <main class="s-main" id="main" role="main">
              
                  
              
              
              
                  
      <section class="s-title-01 mt-lg-5">
          <div class="s-title-01__container container">
              <div class="s-title-01__row row">
                  <div class="s-title-01__col col-lg-8 col-12">
                      <h1 class="s-title-01__title h1">Se connecter | S'inscrire</h1>
                  </div>
              </div>
          </div>
      </section>
      <section class="s-section">
          <div class="s-section__container container">
              <div class="s-section__row row">
                  <div class="s-section__col col-12 col-lg-7">
                      <div class="c-form">
                          <div class="text-center">
                              <a class="proconnect-button" href="/pro_connect/authorize?next="></a>
                              <p>
                                  <a href="https://proconnect.gouv.fr/" rel="noopener noreferrer" target="_blank" title="Qu’est-ce que AgentConnect ? - nouvelle fenêtre">
                                      Qu’est-ce que ProConnect ?
                                  </a>
                              </p>
                          </div>
                          <hr class="my-5" data-it-text="ou"/>
                          <form action="." enctype="multipart/form-data" method="post" novalidate="">
                              <input name="csrfmiddlewaretoken" type="hidden" value="NORMALIZED_CSRF_TOKEN"/>
                              
                              <fieldset>
                                  <p class="h4">Se connecter avec votre email</p>
                                  <p class="text-muted">✨ Nous allons vous envoyer un code magique pour vous connecter sans mot de passe.</p>
                                  
  
  <div class="form-group" id="div_id_email">
      
          
          
              <input class="form-control" id="id_email" maxlength="320" name="email" placeholder="Votre adresse email" required="" type="email"/>
          
      
      
      
      
  </div>
  
                              </fieldset>
                              <div class="row">
                                  <div class="col-12">
                                      <div class="form-row align-items-center justify-content-end gx-3">
                                          <div class="form-group col col-lg-auto order-2 order-lg-3">
                                              <input name="next" type="hidden" value="/topics/"/>
                                              <input class="btn btn-block btn-primary" type="submit" value="Recevoir le lien de connexion"/>
                                          </div>
                                      </div>
                                  </div>
                              </div>
                          </form>
                      </div>
                  </div>
              </div>
          </div>
      </section>
  
              
          </main>
  '''
# ---
# name: TestLoginView.test_content[None][login_view_content]
  '''
  <main class="s-main" id="main" role="main">
              
                  
              
              
              
                  
      <section class="s-title-01 mt-lg-5">
          <div class="s-title-01__container container">
              <div class="s-title-01__row row">
                  <div class="s-title-01__col col-lg-8 col-12">
                      <h1 class="s-title-01__title h1">Se connecter | S'inscrire</h1>
                  </div>
              </div>
          </div>
      </section>
      <section class="s-section">
          <div class="s-section__container container">
              <div class="s-section__row row">
                  <div class="s-section__col col-12 col-lg-7">
                      <div class="c-form">
                          <div class="text-center">
                              <a class="proconnect-button" href="/pro_connect/authorize?next="></a>
                              <p>
                                  <a href="https://proconnect.gouv.fr/" rel="noopener noreferrer" target="_blank" title="Qu’est-ce que AgentConnect ? - nouvelle fenêtre">
                                      Qu’est-ce que ProConnect ?
                                  </a>
                              </p>
                          </div>
                          <hr class="my-5" data-it-text="ou"/>
                          <form action="." enctype="multipart/form-data" method="post" novalidate="">
                              <input name="csrfmiddlewaretoken" type="hidden" value="NORMALIZED_CSRF_TOKEN"/>
                              
                              <fieldset>
                                  <p class="h4">Se connecter avec votre email</p>
                                  <p class="text-muted">✨ Nous allons vous envoyer un code magique pour vous connecter sans mot de passe.</p>
                                  
  
  <div class="form-group" id="div_id_email">
      
          
          
              <input class="form-control" id="id_email" maxlength="320" name="email" placeholder="Votre adresse email" required="" type="email"/>
          
      
      
      
      
  </div>
  
                              </fieldset>
                              <div class="row">
                                  <div class="col-12">
                                      <div class="form-row align-items-center justify-content-end gx-3">
                                          <div class="form-group col col-lg-auto order-2 order-lg-3">
                                              <input name="next" type="hidden" value=""/>
                                              <input class="btn btn-block btn-primary" type="submit" value="Recevoir le lien de connexion"/>
                                          </div>
                                      </div>
                                  </div>
                              </div>
                          </form>
                      </div>
                  </div>
              </div>
          </div>
      </section>
  
              
          </main>
  '''
# ---
# name: TestLoginView.test_post[/-/][login_view_content]
  '''
  <main class="s-main" id="main" role="main">
              
                  
              
              
              
                  
      <section class="s-title-01 mt-lg-5">
          <div class="s-title-01__container container">
              <div class="s-title-01__row row">
                  <div class="s-title-01__col col-lg-8 col-12">
                      <h1 class="s-title-01__title h1">Se connecter | S'inscrire</h1>
                  </div>
              </div>
          </div>
      </section>
      <section class="s-section">
          <div class="s-section__container container">
              <div class="s-section__row row">
                  <div class="s-section__col col-12 col-lg-7">
                      <p>
                          Un lien de connexion vous a été envoyé à l'adresse samuel@jackson.com. Veuillez vérifier votre boîte de réception et cliquer sur le lien pour vous connecter.
                      </p>
                  </div>
              </div>
          </div>
      </section>
  
              
          </main>
  '''
# ---
# name: TestLoginView.test_post[/topics/-/topics/][login_view_content]
  '''
  <main class="s-main" id="main" role="main">
              
                  
              
              
              
                  
      <section class="s-title-01 mt-lg-5">
          <div class="s-title-01__container container">
              <div class="s-title-01__row row">
                  <div class="s-title-01__col col-lg-8 col-12">
                      <h1 class="s-title-01__title h1">Se connecter | S'inscrire</h1>
                  </div>
              </div>
          </div>
      </section>
      <section class="s-section">
          <div class="s-section__container container">
              <div class="s-section__row row">
                  <div class="s-section__col col-12 col-lg-7">
                      <p>
                          Un lien de connexion vous a été envoyé à l'adresse samuel@jackson.com. Veuillez vérifier votre boîte de réception et cliquer sur le lien pour vous connecter.
                      </p>
                  </div>
              </div>
          </div>
      </section>
  
              
          </main>
  '''
# ---
# name: TestLoginView.test_post[http://www.unallowed_host.com-/][login_view_content]
  '''
  <main class="s-main" id="main" role="main">
              
                  
              
              
              
                  
      <section class="s-title-01 mt-lg-5">
          <div class="s-title-01__container container">
              <div class="s-title-01__row row">
                  <div class="s-title-01__col col-lg-8 col-12">
                      <h1 class="s-title-01__title h1">Se connecter | S'inscrire</h1>
                  </div>
              </div>
          </div>
      </section>
      <section class="s-section">
          <div class="s-section__container container">
              <div class="s-section__row row">
                  <div class="s-section__col col-12 col-lg-7">
                      <p>
                          Un lien de connexion vous a été envoyé à l'adresse samuel@jackson.com. Veuillez vérifier votre boîte de réception et cliquer sur le lien pour vous connecter.
                      </p>
                  </div>
              </div>
          </div>
      </section>
  
              
          </main>
  '''
# ---
# name: TestSendMagicLink.test_send_magic_link[DEV-1][send_magic_link_payload]
  '{"sender": {"name": "La Communaut\\u00e9", "email": "noreply@inclusion.gouv.fr"}, "to": [{"email": "samuel@jackson.com"}], "params": {"display_name": "Samuel J.", "login_link": "LOGIN_LINK"}, "templateId": 31}'
# ---
# name: TestSendMagicLink.test_send_magic_link[PROD-0][send_magic_link_payload]
  '{"sender": {"name": "La Communaut\\u00e9", "email": "noreply@inclusion.gouv.fr"}, "to": [{"email": "samuel@jackson.com"}], "params": {"display_name": "Samuel J.", "login_link": "LOGIN_LINK"}, "templateId": 31}'
# ---
